extension/src/runners/runnerExecutor.ts

```
@@ -0,0 +1,154 @@
+    private readonly javaHome: string;
+    private readonly context: ExtensionContext;
+    private isRunning: boolean;
+    private preLaunchTask: cp.ChildProcess | undefined;
+    private runnerMap: Map<ITestRunner, ITestItem[]> | undefined;
+
+    constructor(javaHome: string, context: ExtensionContext) {
+        this.javaHome = javaHome;
+        this.context = context;
+    }
+
+    public async run(testItems: ITestItem[], isDebug: boolean = false, config?: IExecutionConfig): Promise<void> {
+        if (this.isRunning) {
+            window.showInformationMessage('A test session is currently running. Please wait until it finishes.');
+            return;
+        }
+        this.isRunning = true;
+        testStatusBarProvider.showRunningTest();
+        try {
+            this.runnerMap = this.classifyTestsByKind(testItems);
+            for (const [runner, tests] of this.runnerMap.entries()) {
+                if (config && config.preLaunchTask.length > 0) {
+                    this.preLaunchTask = cp.exec(
+                        config.preLaunchTask,
+                        {
+                            maxBuffer: CHILD_PROCESS_MAX_BUFFER_SIZE,
+                            cwd: config.workingDirectory,
+                        },
+                    );
+                    await this.execPreLaunchTask();
+                }
+                await runner.setup(tests, isDebug, config);
+                const results: ITestResult[] = await runner.run();
+                testResultManager.storeResult(...results);
+                testStatusBarProvider.showTestResult(results);
+                testCodeLensProvider.refresh();
+            }
+        } catch (error) {
+            window.showErrorMessage(`${error}`);
+        } finally {
+            await this.cleanUp();
+        }
+    }
+
+    private async cleanUp(): Promise<void> {
+        try {
+            if (this.preLaunchTask) {
+                await killProcess(this.preLaunchTask);
```

What happens if the preLaunchTask is already closed?

---

Changed the logic: swallow the error when we kill the process